servers/services/src/main/java/org/projectnessie/services/impl/NamespaceApiImpl.java

```diff
@@ -0,0 +1,279 @@
+    Optional<Content> explicitlyCreatedNamespace = getExplicitlyCreatedNamespace(params, branch);
+    if (explicitlyCreatedNamespace.isPresent()) {
+      return explicitlyCreatedNamespace
+          .get()
+          .unwrap(Namespace.class)
+          .orElseThrow(() -> namespaceDoesNotExistException(params));
+    }
+
+    return getImplicitlyCreatedNamespace(params, branch)
+        .orElseThrow(() -> namespaceDoesNotExistException(params));
+  }
+
+  @Override
+  public GetNamespacesResponse getNamespaces(NamespaceParams params)
+      throws NessieReferenceNotFoundException {
+    BranchName branch = branchFromRefName(params);
+    try {
+      Set<Namespace> allNamespaces =
+          Sets.newHashSet(getExplicitlyCreatedNamespaces(params, branch));
+      List<Namespace> implicitlyCreatedNamespaces = getImplicitlyCreatedNamespaces(params, branch);
+      allNamespaces.addAll(implicitlyCreatedNamespaces);
+      return ImmutableGetNamespacesResponse.builder().addAllNamespaces(allNamespaces).build();
+    } catch (ReferenceNotFoundException e) {
+      throw refNotFoundException(e);
+    }
+  }
+
+  /**
+   * Retrieves a list of explicitly created namespaces (via {@link
+   * NamespaceApi#createNamespace(NamespaceParams)}).
+   *
+   * @param params The params to use for fetching the namespaces
+   * @param branch The ref
+   * @return A list of explicitly created namespaces.
+   * @throws ReferenceNotFoundException If the ref could not be found.
+   */
+  private List<Namespace> getExplicitlyCreatedNamespaces(NamespaceParams params, BranchName branch)
+      throws ReferenceNotFoundException {
+    try (Stream<Key> keyStream =
+        getStore()
+            .getKeys(branch)
+            .filter(k -> Type.NAMESPACE == k.getType())
+            .filter(
+                k ->
+                    Namespace.of(k.getValue().getElements())
+                        .name()
+                        .startsWith(params.getNamespace().name()))
+            .map(WithType::getValue)) {
+
+      List<Key> keys = keyStream.collect(Collectors.toList());
+      Map<Key, Content> values = getStore().getValues(branch, keys);
+      return values.values().stream()
+          .map(c -> c.unwrap(Namespace.class).get())
+          .collect(Collectors.toList());
+    }
+  }
+
+  /**
+   * Retrieves a list of implicit namespaces that might contain duplicate entries. An implicitly
+   * created namespace generally occurs when adding a table 'a.b.c.table' where the 'a.b.c' part
+   * represents the namespace.
+   *
+   * @param params The params to use for fetching the namespaces
+   * @param branch The ref
+   * @return A list of implicit namespaces that might contain duplicate entries.
+   */
+  private List<Namespace> getImplicitlyCreatedNamespaces(NamespaceParams params, BranchName branch)
+      throws ReferenceNotFoundException {
+    try (Stream<WithType<Key, Type>> stream = getImplicitNamespacesStream(params, branch)) {
```

Comment:
Checks for explicit and implicit namespaces should be unified, because the stream returning the implicit ones includes the explicitly created one.